File: 010 栈的基本实现/使用顺序表实现.swift
import Foundation

struct Stack<E> {
    var top: Int
    var maxsize: Int
    var elems: [E?]
    
    init(_ maxsize: Int = 20) {
        self.maxsize = maxsize
        self.elems = Array<E?>(repeating: nil, count: maxsize)
        self.top = 0
    }
    
    mutating func clear() {
        top = 0
    }
    
    func isEmpty() -> Bool {
        return top == 0
    }
    
    func length() -> Int {
        return top
    }
    
    func getTop() -> E? {
        return top > 0 ? elems[top - 1] : nil
    }
    
    mutating func push(_ e: E) -> Bool {
        if top >= maxsize {
            return false
        }
        elems[top] = e
        top += 1
        return true
    }
    
    mutating func pop() -> E? {
        if top > 0 {
            let e = elems[top - 1]
            top -= 1
            return e
        }
        return nil
    }
    
    func traverse() {
        for i in 0..<top {
            print(elems[i]!, terminator: " ")
        }
        print()
    }
}

var stack = Stack<Int>()

for _ in 0..<20 {
    let v = Int.random(in: 0..<100)
    print(v, terminator: " ")
    stack.push(v)
}
print()

stack.traverse()

print(stack.getTop())
print(stack.isEmpty())
print(stack.length())

print(stack.pop())
stack.traverse()
print(stack.getTop())
print(stack.isEmpty())
print(stack.length())


stack.clear()
stack.traverse()
print(stack.isEmpty())
print(stack.length())

print(stack.pop())
stack.traverse()
print(stack.getTop())
print(stack.isEmpty())
print(stack.length())

File: 010 栈的基本实现/使用顺序表实现.c
#include <stdio.h>
#include <stdlib.h>

typedef struct {
    int size;
    int *base;
    int top;
} Stack;

Stack * createStack(int size) {
    Stack *stack = (Stack *)malloc(sizeof(Stack));
    stack->size = size;
    stack->base = malloc(sizeof(int) * size);
    stack->top = 0;
    return stack;
}

void clean(Stack *stack) {
    stack->top = 0;
}

int length(Stack *stack) {
    return stack->top;
}

int getTop(Stack *stack) {
    if (stack->top == 0) {
        return -1000;
    }
    int idx = stack->top - 1;
    return stack->base[idx];
}

void push(Stack *stack, int e) {
    if (stack->top == stack->size) {
        return;
    }
    stack->base[stack->top] = e;
    stack->top += 1;
}

int pop(Stack *stack) {
    if (stack->top == 0) {
        return -10000;
    }
    stack->top -= 1;
    return stack->base[stack->top];
}

int isEmpty(Stack *stack) {
    return stack->top == 0;
}

void traverse(Stack *stack) {
    for (int i = 0; i < stack->top; i++) {
        printf("%d ", stack->base[i]);
    }
    printf("\n");
}

int main() {
    Stack *stack = createStack(100);
    for (int i = 0; i < 20; i++) {
        int v = rand() % 100;
        push(stack, v);
        printf("%d ", v);
    }
    printf("\n");
    
    traverse(stack);
    
    printf("%d \n", getTop(stack));
    printf("%d \n", isEmpty(stack));
    printf("%d \n", length(stack));
    
    printf("%d \n", pop(stack));
    traverse(stack);
    printf("%d \n", getTop(stack));
    printf("%d \n", isEmpty(stack));
    printf("%d \n", length(stack));
    
    clean(stack);
    traverse(stack);
    printf("%d \n", getTop(stack));
    printf("%d \n", isEmpty(stack));
    printf("%d \n", length(stack));
    
    pop(stack);
    traverse(stack);
    printf("%d \n", getTop(stack));
    printf("%d \n", isEmpty(stack));
    printf("%d \n", length(stack));
    return 0;
}


